atc/db/resource.go

```diff
@@ -949,3 +993,263 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
 
 	return nil
 }
+
+// this allows us to reuse getCausalityResourceVersions to construct both upstream and downstream trees by passing in a different updater fn
+type resourceVersionUpdater func(*atc.CausalityResourceVersion, *atc.CausalityBuild)
+
+func contains(lst []int, e int) bool {
+	for _, v := range lst {
+		if e == v {
+			return true
+		}
+	}
+	return false
+}
+
+func causalityQuery(direction CausalityDirection) string {
+	switch direction {
```

Comment:
I notice you deployed the updated version to CI. Performance doesn't seem too bad with the version_md5 based approach!

I haven't tried to reason about why this is happening yet (i.e. whether it's a bug with the UI or the API endpoints), but it seems like the graph is rendered in the wrong direction - e.g. when looking at the downstream graph, the resource version appears on the right of the graph with the job edges feeding into it, rather than out of it (e.g. https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/concourse/causality/999241219/downstream):

<img width="1277" alt="Screen Shot 2021-06-25 at 3 32 06 PM" src="https://user-images.githubusercontent.com/26583442/123476300-82d46600-d5ca-11eb-9faf-9279d05f2e8b.png">

Reply:
This, along with some other subtle bugs were fixed in https://github.com/concourse/concourse/pull/7125/commits/ff4a1a0c750471907aab693928b72630c029d5df